### PR TITLE
Functional tests refactored (closes #356)

### LINKED ISSUE #356: I have some questions regarding our functional testing
Ok, I was not quite attentive to what's going on with our functional testing. Looking at it currently I have some questions:

1. https://github.com/DevExpress/testcafe/blob/master/test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js#L35 - why `runTests` resolves with error?
2. https://github.com/DevExpress/testcafe/blob/master/test/functional/fixtures/runner/uncaught-js-errors/uncaught-js-errors-test.js#L47 - Why tests that names starts with `Should fail` expects successful completion?
3. https://github.com/DevExpress/testcafe/tree/master/test/functional/fixtures/runner/api/click - it's hard to determine the actual test and testcafe-test, the difference with naming is minimal. Can we have subdirs, e.g. `quarantine-mode/test.js` - actual test, `quarantine-mode/pages/*.html` - markup, `quarantine-mode/testcafe-fixtures/*.test.js` - TC fixtures?
4.  Why quarantine mode tests are not inside `runner` directory? BTW, I don't think we need `runner` directory` for now. We can start with the plain dir structure for now.
5. https://github.com/DevExpress/testcafe/blob/master/test/functional/fixtures/quarantine-mode/index-test.js#L40 - why we need additional server here? Can't we incorporate this functionality to the already existing test server?

\cc @AlexanderMoskovkin @georgiy-abbasov @AndreyBelym

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @AlexanderMoskovkin, @VasilyStrelyaev, @inikulin

### CODE REVIEW

File: test/functional/setup.js [hunk 1]
```diff
@@ -126,10 +126,11 @@ before(function () {
 
                         global.testReport = testReport;
 
-                        if (shouldFail)
-                            throw testError;
+                        if (shouldFail && !testError)
+                            throw new Error('Test should be failed but it is succeed');
```
Comment (VasilyStrelyaev) [inline test/functional/setup.js]:
```
                            throw new Error('Test should fail but it succeed');
```

File: test/functional/fixtures/using-shared-code/test.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    });
+
+    it('Should pass if using a mixin in a same-domain page', function () {
+        return runTests(TC_FIXTURES_DIR + 'shared-code.test.js', 'Using mixin in iframe   →   Same-domain iframe');
+    });
+
+    it('Should pass if using a mixin in a iframe page without source', function () {
+        return runTests(TC_FIXTURES_DIR + 'shared-code.test.js', 'Using mixin in iframe   →   Inline iframe');
+    });
+});
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/using-shared-code/test.js]: here it's better to use *when* instead of *if*
and *an iframe*

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

File: test/functional/fixtures/api/click/test.js [hunk 3]
```diff
@@ -1,8 +1,11 @@
 var expect = require('chai').expect;
 
+var TC_FIXTURES_DIR = 'testcafe-fixtures/';
```
Comment (inikulin) [inline test/functional/fixtures/api/click/test.js]: I don't see any reason why we need this constant

File: test/functional/fixtures/api/click/test_config.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (inikulin) [inline test/functional/fixtures/api/click/test_config.json]: `test_config.json` belong to testcafe tests. They should be in `testcafe-fixtures` dir

File: test/functional/fixtures/using-shared-code/testcafe-fixtures/shared-code.test.js [hunk 5]
```diff
@@ -2,10 +2,10 @@
 '@page ./index.html';
 
 '@require :testModule';
-'@require ./mixin.js';
+'@require ../mixins/mixin.js';
```
Comment (inikulin) [inline test/functional/fixtures/using-shared-code/testcafe-fixtures/shared-code.test.js]: No need to put mixins and helpers in the separate dir. Also they belong to testcafe tests, so they should be in `testcafe-fixtures` dir as `mixins.js` and `helpers.js` respectively.

File: test/functional/fixtures/api/click/test.js (hunk 3 above)
Comment (inikulin) [inline test/functional/fixtures/api/click/test.js]: Just use full path inline

**Comment by inikulin** [comment]
\r-

**Comment by georgiy-abbasov** [comment]
FPR

**Comment by inikulin** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Pull request merged**